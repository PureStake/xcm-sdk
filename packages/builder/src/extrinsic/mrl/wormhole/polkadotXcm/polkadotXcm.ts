import { getMultilocationDerivedAddresses } from '@moonbeam-network/xcm-utils';
import { AssetAmount } from '@moonbeam-network/xcm-types';
import { ExtrinsicBuilder } from '../../../ExtrinsicBuilder';
import { MrlExtrinsicConfigBuilder } from '../../../ExtrinsicBuilder.interfaces';
import { ExtrinsicConfig } from '../../../ExtrinsicConfig';

// TODO: Can we move them somewhere?
const BUY_EXECUTION_FEE = 100_000_000_000_000_000n;
const CROSS_CHAIN_FEE = 100_000_000_000_000_000n;
export const BATCH_CONTRACT_ADDRESS =
  '0x0000000000000000000000000000000000000808';

export function polkadotXcm() {
  return {
    send: (): MrlExtrinsicConfigBuilder => ({
      build: (params) => {
        const {
          destination,
          destinationAddress,
          moonAsset,
          moonChain,
          sourceApi,
          transact,
        } = params;

        if (!destination.wh?.name) {
          throw new Error('Destination chain does not have a wormhole name');
        }

        if (!transact) {
          throw new Error('Transact params are required');
        }

        const { transfer } = sourceApi.tx.xTokens;
        const builder = ExtrinsicBuilder().xTokens().transfer();

        const assetTx = transfer(builder.build(params).getArgs(transfer));
        /*
         * TODO: Can we move it to AssetRoute and receive it in build params?
         * In the future we could also check if wallet already has necessary DEV/GLMR to pay execution fees on Moonbase/Moonbeam.
         * Also we need to move fees to AssetRoute.
         */
        const executionAssetTx = transfer(
          builder
            .build({
              ...params,
              asset: AssetAmount.fromChainAsset(moonAsset, {
                amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
              }),
            })
            .getArgs(transfer),
        );

        const { address20 } = getMultilocationDerivedAddresses({
          address: destinationAddress,
          paraId: moonChain.parachainId,
          isParents: true,
        });

        const send = sourceApi.tx.polkadotXcm.send(
          {
            V3: {
              parents: 1,
              interior: { X1: { Parachain: moonChain.parachainId } },
            },
          },
          {
            V3: [
              {
                WithdrawAsset: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                ],
              },
              {
                BuyExecution: {
                  fees: {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                  weightLimit: 'Unlimited',
                },
              },
              {
                Transact: {
                  originKind: 'SovereignAccount',
                  requireWeightAtMost: {
                    refTime: transact.txWeight.refTime,
                    proofSize: transact.txWeight.proofSize,
                  },
                  call: {
                    encoded: transact.call,
                  },
                },
              },
              {
                SetAppendix: [
                  { RefundSurplus: {} },
                  {
                    DepositAsset: {
                      assets: { Wild: { AllCounted: 1 } },
                      beneficiary: {
                        parents: 0,
                        interior: {
                          X1: { AccountKey20: { key: address20 } },
                        },
                      },
                    },
                  },
                ],
              },
            ],
          },
        );

        return new ExtrinsicConfig({
          module: 'utility',
          func: 'batchAll',
          getArgs: () => [assetTx, executionAssetTx, send],
        });
      },
    }),
  };
}
