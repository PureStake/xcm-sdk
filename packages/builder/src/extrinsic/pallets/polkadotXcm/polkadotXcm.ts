import { ExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import { ExtrinsicConfig } from '../../ExtrinsicConfig';
import {
  getPolkadotXcmExtrinsicArgs,
  shouldFeeAssetPrecedeAsset,
} from './polkadotXcm.util';

const pallet = 'polkadotXcm';

export function polkadotXcm() {
  return {
    limitedReserveTransferAssets: () => {
      const func = 'limitedReserveTransferAssets';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) =>
                getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: {
                        Concrete: {
                          parents: 0,
                          interior: 'Here',
                        },
                      },
                      fun: {
                        Fungible: params.amount,
                      },
                    },
                  ],
                }),
            }),
        }),
        X1: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) =>
                getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: {
                        Concrete: {
                          parents: 0,
                          interior: {
                            X1: {
                              PalletInstance: params.palletInstance,
                            },
                          },
                        },
                      },
                      fun: {
                        Fungible: params.amount,
                      },
                    },
                  ],
                }),
            }),
        }),
        X2: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const isAssetDifferent =
                  !!params.feeAsset && params.asset !== params.feeAsset;
                const assets = [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X2: [
                            {
                              PalletInstance: params.palletInstance,
                            },
                            {
                              GeneralIndex: params.asset,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: params.amount,
                    },
                  },
                ];

                const shouldFeeAssetPrecede =
                  shouldFeeAssetPrecedeAsset(params);

                if (isAssetDifferent) {
                  const feeAsset = {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X2: [
                            {
                              PalletInstance: params.palletInstance,
                            },
                            {
                              GeneralIndex: params.feeAsset,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: params.fee,
                    },
                  };

                  if (shouldFeeAssetPrecede) {
                    assets.unshift(feeAsset);
                  } else {
                    assets.push(feeAsset);
                  }
                }

                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: assets,
                  feeIndex: isAssetDifferent && !shouldFeeAssetPrecede ? 1 : 0,
                });
              },
            }),
        }),
      };
    },
    limitedReserveWithdrawAssets: () => {
      const func = 'limitedReserveWithdrawAssets';

      return {
        X2: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) =>
                getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: {
                        Concrete: {
                          parents: 1,
                          interior: {
                            X2: [
                              {
                                Parachain: params.destination.parachainId,
                              },
                              {
                                PalletInstance: params.palletInstance,
                              },
                            ],
                          },
                        },
                      },
                      fun: {
                        Fungible: params.amount,
                      },
                    },
                  ],
                }),
            }),
        }),
      };
    },
  };
}
