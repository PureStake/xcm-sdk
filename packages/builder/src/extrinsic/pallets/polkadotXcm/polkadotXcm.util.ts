/* eslint-disable sort-keys */
import { SubmittableExtrinsicFunction } from '@polkadot/api/types';
import {
  ExtrinsicConfigBuilderPrams,
  Parents,
} from '../../ExtrinsicBuilder.interfaces';
import {
  getExtrinsicAccount,
  getExtrinsicArgumentVersion,
} from '../../ExtrinsicBuilder.utils';

export interface GetExtrinsicParams extends ExtrinsicConfigBuilderPrams {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  asset: any;
  func?: SubmittableExtrinsicFunction<'promise'>;
  parents?: Parents;
  feeIndex?: number;
}

export function getPolkadotXcmExtrinsicArgs({
  asset,
  address,
  destination,
  func,
  parents = 1,
  feeIndex = 0,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}: GetExtrinsicParams): any[] {
  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  const version = getExtrinsicArgumentVersion(func);

  return [
    {
      [version]: {
        parents,
        interior: {
          X1: {
            Parachain: destination.parachainId,
          },
        },
      },
    },
    {
      [version]: {
        parents: 0,
        interior: {
          X1: getExtrinsicAccount(address),
        },
      },
    },
    {
      [version]: asset,
    },
    feeIndex,
    'Unlimited',
  ];
}

export function shouldFeeAssetPrecedeAsset({
  asset,
  feeAsset,
}: ExtrinsicConfigBuilderPrams): boolean {
  const assetIdNumber = Number(asset);
  const feeAssetIdNumber = Number(feeAsset);

  if (Number.isNaN(assetIdNumber) || Number.isNaN(feeAssetIdNumber)) {
    return false;
  }

  return assetIdNumber > feeAssetIdNumber;
}
