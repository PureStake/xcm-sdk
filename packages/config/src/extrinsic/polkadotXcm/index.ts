export * from './polkadotXcm';
export * from './polkadotXcm.interfaces';
