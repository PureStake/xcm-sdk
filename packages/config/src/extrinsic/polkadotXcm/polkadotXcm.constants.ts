export enum PolkadotXcmExtrinsic {
  LimitedReserveTransferAssets = 'limitedReserveTransferAssets',
  LimitedReserveWithdrawAssets = 'limitedReserveWithdrawAssets',
}

export enum PolkadotXcmExtrinsicSuccessEvent {
  Attempted = 'Attempted',
}
