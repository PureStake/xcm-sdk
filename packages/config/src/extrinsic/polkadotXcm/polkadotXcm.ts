import { AssetConfig, ChainConfig, MoonChainConfig } from '../../interfaces';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';
import { getCreateExtrinsic } from './polkadotXcm.util';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function polkadotXcm<Assets>(chain: MoonChainConfig) {
  return {
    limitedReserveTransferAssets: () =>
      limitedReserveTransferAssets<Assets>(chain),
    limitedReserveWithdrawAssets: () => limitedReserveWithdrawAssets(chain),
  };
}

function limitedReserveTransferAssets<Assets>(chain: MoonChainConfig) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveTransferAssets,
          event,
          chain,
          origin,
        );

        return {
          V0: () =>
            createExtrinsic((amount) => ({
              V0: [
                {
                  ConcreteFungible: {
                    id: 'Null',
                    amount,
                  },
                },
              ],
            })),
          V1: () => ({
            here: () =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: 'Here',
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
            X1: () =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: 5,
                          },
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
            X2: (palletInstance: number, asset: AssetConfig<Assets>) =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X2: [
                            {
                              PalletInstance: palletInstance,
                            },
                            {
                              // TODO: throw an error if id is not defined or fix types
                              GeneralIndex: asset.originAssetId || 0,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
          }),
        };
      },
    }),
  };
}

function limitedReserveWithdrawAssets(chain: MoonChainConfig) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveWithdrawAssets,
          event,
          chain,
          origin,
        );

        return {
          V1: () => ({
            X2: (palletInstance: number) =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 1,
                        interior: {
                          X2: [
                            {
                              PalletInstance: palletInstance,
                            },
                            {
                              Parachain: chain.parachainId,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
          }),
        };
      },
    }),
  };
}
