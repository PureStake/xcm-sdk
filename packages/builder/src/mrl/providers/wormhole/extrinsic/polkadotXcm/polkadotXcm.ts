import { type AnyParachain, AssetAmount } from '@moonbeam-network/xcm-types';
import { getMultilocationDerivedAddresses } from '@moonbeam-network/xcm-utils';
import { ExtrinsicBuilder } from '../../../../../extrinsic/ExtrinsicBuilder';
import { ExtrinsicConfig } from '../../../../../types/substrate/ExtrinsicConfig';
import type { MrlConfigBuilder } from '../../../../MrlBuilder.interfaces';

// TODO: Can we move them somewhere?
const BUY_EXECUTION_FEE = 100_000_000_000_000_000n;
const CROSS_CHAIN_FEE = 100_000_000_000_000_000n;

export function polkadotXcm() {
  return {
    send: (): MrlConfigBuilder => ({
      build: ({
        asset,
        destination,
        destinationAddress,
        destinationApi,
        fee,
        moonAsset,
        moonChain,
        moonApi,
        source,
        sourceAddress,
        sourceApi,
        transact,
      }) => {
        if (!destination.wh?.name) {
          throw new Error('Destination chain does not have a wormhole name');
        }

        if (!transact) {
          throw new Error('Transact params are required');
        }

        if (!sourceApi) {
          throw new Error('Source API needs to be defined');
        }

        const { address20: computedOriginAccount } =
          getMultilocationDerivedAddresses({
            address: sourceAddress,
            paraId: moonChain.parachainId,
            isParents: true,
          });

        const { transfer } = sourceApi.tx.xTokens;
        const builder = ExtrinsicBuilder().xTokens().transfer();

        const assetTransferTx = transfer(
          ...builder
            .build({
              asset,
              destination: moonChain,
              destinationAddress: computedOriginAccount,
              destinationApi: moonApi,
              fee,
              // TODO: This is a workaround. xTokens.transfer doesn't need source chain but the interfaces requires it.
              // In this case we know that a source chain is not a Parachain.
              source: source as AnyParachain,
              sourceAddress,
              sourceApi,
            })
            .getArgs(transfer),
        );
        /*
         * TODO: Can we move it to AssetRoute and receive it in build params?
         * In the future we could also check if wallet already has necessary DEV/GLMR to pay execution fees on Moonbase/Moonbeam.
         * Also we need to move fees to AssetRoute.
         */
        const feeAssetTransferTx = transfer(
          ...builder
            .build({
              asset: AssetAmount.fromChainAsset(moonAsset, {
                amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
              }),
              destination: moonChain,
              destinationAddress,
              destinationApi: moonApi,
              fee,
              source: source as AnyParachain,
              sourceAddress,
              sourceApi,
            })
            .getArgs(transfer),
        );

        const send = sourceApi.tx.polkadotXcm.send(
          {
            V3: {
              parents: 1,
              interior: { X1: { Parachain: moonChain.parachainId } },
            },
          },
          {
            V3: [
              {
                WithdrawAsset: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                ],
              },
              {
                BuyExecution: {
                  fees: {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                  weightLimit: 'Unlimited',
                },
              },
              {
                Transact: {
                  originKind: 'SovereignAccount',
                  requireWeightAtMost: {
                    refTime: transact.txWeight.refTime,
                    proofSize: transact.txWeight.proofSize,
                  },
                  call: {
                    encoded: transact.call,
                  },
                },
              },
              {
                SetAppendix: [
                  { RefundSurplus: {} },
                  {
                    DepositAsset: {
                      assets: { Wild: { AllCounted: 1 } },
                      beneficiary: {
                        parents: 0,
                        interior: {
                          X1: { AccountKey20: { key: computedOriginAccount } },
                        },
                      },
                    },
                  },
                ],
              },
            ],
          },
        );

        // TODO add here ability to only send the remote execution (only `send`)
        return new ExtrinsicConfig({
          module: 'utility',
          func: 'batchAll',
          getArgs: () => [[assetTransferTx, feeAssetTransferTx, send]],
        });
      },
    }),
  };
}
