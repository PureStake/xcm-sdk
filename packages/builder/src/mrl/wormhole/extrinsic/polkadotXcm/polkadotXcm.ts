import { getMultilocationDerivedAddresses } from '@moonbeam-network/xcm-utils';
import {
  AnyParachain,
  AssetAmount,
  EvmParachain,
  Parachain,
} from '@moonbeam-network/xcm-types';
import { ExtrinsicBuilder } from '../../../../extrinsic/ExtrinsicBuilder';
import { ExtrinsicConfig } from '../../../../types/substrate/ExtrinsicConfig';
import { MrlConfigBuilder } from '../../../MrlBuilder.interfaces';

// TODO: Can we move them somewhere?
const BUY_EXECUTION_FEE = 100_000_000_000_000_000n;
const CROSS_CHAIN_FEE = 100_000_000_000_000_000n;
export const BATCH_CONTRACT_ADDRESS =
  '0x0000000000000000000000000000000000000808';

export function polkadotXcm() {
  return {
    send: (): MrlConfigBuilder => ({
      build: ({
        asset,
        destination,
        destinationAddress,
        destinationApi,
        fee,
        moonAsset,
        moonChain,
        source,
        sourceAddress,
        sourceApi,
        transact,
      }) => {
        if (!destination.wh?.name) {
          throw new Error('Destination chain does not have a wormhole name');
        }

        if (!Parachain.is(destination) && !EvmParachain.is(destination)) {
          throw new Error(
            `Destination ${destination.name} is not a Parachain or EvmParachain`,
          );
        }

        if (!transact) {
          throw new Error('Transact params are required');
        }

        const { transfer } = sourceApi.tx.xTokens;
        const builder = ExtrinsicBuilder().xTokens().transfer();

        const assetTransferTx = transfer(
          builder
            .build({
              asset,
              destination,
              destinationAddress,
              destinationApi,
              fee,
              // TODO: This is a workaround. xTokens.transfer doesn't need source chain but the interfaces requires it.
              // In this case we know that a source chain is not a Parachain.
              source: source as AnyParachain,
              sourceAddress,
              sourceApi,
            })
            .getArgs(transfer),
        );
        /*
         * TODO: Can we move it to AssetRoute and receive it in build params?
         * In the future we could also check if wallet already has necessary DEV/GLMR to pay execution fees on Moonbase/Moonbeam.
         * Also we need to move fees to AssetRoute.
         */
        const feeAssetTransferTx = transfer(
          builder
            .build({
              asset: AssetAmount.fromChainAsset(moonAsset, {
                amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
              }),
              destination,
              destinationAddress,
              destinationApi,
              fee,
              source: source as AnyParachain,
              sourceAddress,
              sourceApi,
            })
            .getArgs(transfer),
        );

        const { address20 } = getMultilocationDerivedAddresses({
          address: destinationAddress,
          paraId: moonChain.parachainId,
          isParents: true,
        });

        const send = sourceApi.tx.polkadotXcm.send(
          {
            V3: {
              parents: 1,
              interior: { X1: { Parachain: moonChain.parachainId } },
            },
          },
          {
            V3: [
              {
                WithdrawAsset: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                ],
              },
              {
                BuyExecution: {
                  fees: {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: moonAsset.getAssetPalletInstance(),
                          },
                        },
                      },
                    },
                    fun: { Fungible: BUY_EXECUTION_FEE },
                  },
                  weightLimit: 'Unlimited',
                },
              },
              {
                Transact: {
                  originKind: 'SovereignAccount',
                  requireWeightAtMost: {
                    refTime: transact.txWeight.refTime,
                    proofSize: transact.txWeight.proofSize,
                  },
                  call: {
                    encoded: transact.call,
                  },
                },
              },
              {
                SetAppendix: [
                  { RefundSurplus: {} },
                  {
                    DepositAsset: {
                      assets: { Wild: { AllCounted: 1 } },
                      beneficiary: {
                        parents: 0,
                        interior: {
                          X1: { AccountKey20: { key: address20 } },
                        },
                      },
                    },
                  },
                ],
              },
            ],
          },
        );

        return new ExtrinsicConfig({
          module: 'utility',
          func: 'batchAll',
          getArgs: () => [assetTransferTx, feeAssetTransferTx, send],
        });
      },
    }),
  };
}
