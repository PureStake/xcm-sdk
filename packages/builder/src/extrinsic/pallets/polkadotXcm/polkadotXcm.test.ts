import { buildParamsMock } from '../../../../fixtures';
import { polkadotXcm } from './polkadotXcm';

describe('polkadotXcm', () => {
  describe('limitedReserveTransferAssets', () => {
    describe('here', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .here()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x1', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X1()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x2', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X2()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('limitedReserveWithdrawAssets', () => {
    describe('x2', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveWithdrawAssets()
        .X2()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });
});
