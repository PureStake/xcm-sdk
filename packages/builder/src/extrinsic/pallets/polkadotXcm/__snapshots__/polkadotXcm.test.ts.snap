// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polkadotXcm limitedReserveTransferAssets here should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "pallet": "polkadotXcm",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets here should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveTransferAssets x1 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "pallet": "polkadotXcm",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets x1 should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X1": {
                "PalletInstance": 5,
              },
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveTransferAssets x2 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "pallet": "polkadotXcm",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets x2 should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveWithdrawAssets x2 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveWithdrawAssets",
  "getArgs": [Function],
  "pallet": "polkadotXcm",
}
`;

exports[`polkadotXcm limitedReserveWithdrawAssets x2 should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 1000,
                },
                {
                  "PalletInstance": 10,
                },
              ],
            },
            "parents": 1,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;
